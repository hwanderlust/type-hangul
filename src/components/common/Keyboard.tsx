import React, { useState } from "react";
import styled from "styled-components";

import { Fonts, KeyType, Sizes, StyledProps } from "../../helpers";

const topRow: Array<KeyType> = [
  { eng: "q", shift: "ㅃ", kor: "ㅂ" },
  { eng: "w", shift: "ㅉ", kor: "ㅈ" },
  { eng: "e", shift: "ㄸ", kor: "ㄷ" },
  { eng: "r", shift: "ㄲ", kor: "ㄱ" },
  { eng: "t", shift: "ㅆ", kor: "ㅅ" },
  { eng: "y", shift: "", kor: "ㅛ" },
  { eng: "u", shift: "", kor: "ㅕ" },
  { eng: "i", shift: "", kor: "ㅑ" },
  { eng: "o", shift: "ㅒ", kor: "ㅐ" },
  { eng: "p", shift: "ㅖ", kor: "ㅔ" },
];
const midRow: Array<KeyType> = [
  { eng: "a", shift: "", kor: "ㅁ" },
  { eng: "s", shift: "", kor: "ㄴ" },
  { eng: "d", shift: "", kor: "ㅇ" },
  { eng: "f", shift: "", kor: "ㄹ" },
  { eng: "g", shift: "", kor: "ㅎ" },
  { eng: "h", shift: "", kor: "ㅗ" },
  { eng: "j", shift: "", kor: "ㅓ" },
  { eng: "k", shift: "", kor: "ㅏ" },
  { eng: "l", shift: "", kor: "ㅣ" },
];
const botRow: Array<KeyType> = [
  { eng: "z", shift: "", kor: "ㅋ" },
  { eng: "x", shift: "", kor: "ㅌ" },
  { eng: "c", shift: "", kor: "ㅊ" },
  { eng: "v", shift: "", kor: "ㅍ" },
  { eng: "b", shift: "", kor: "ㅠ" },
  { eng: "n", shift: "", kor: "ㅜ" },
  { eng: "m", shift: "", kor: "ㅡ" },
];

const Container = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  align-items: center;
`;
interface RowProps {
  numOfEl: number;
}
const Row = styled.div<RowProps>`
  display: inline-flex;
  justify-content: space-between;
  margin-bottom: calc(4px + (8 - 4) * ((100vw - 300px) / (1440 - 300)));

  &:last-of-type {
    margin-bottom: 0;
  }
`;
const Key = styled.div`
  background-color: black;
  width: calc(25px + (75 - 25) * ((100vw - 300px) / (1440 - 300)));
  height: calc(40px + (75 - 40) * ((100vw - 300px) / (1440 - 300)));
  display: flex;
  justify-content: center;
  align-items: center;
  border-top-left-radius: 20px;
  border-bottom-left-radius: 20px;
  border-top-right-radius: 20px;
  border-bottom-right-radius: 20px;
  margin-right: calc(4px + (8 - 4) * ((100vw - 300px) / (1440 - 300)));

  &:last-of-type {
    margin-right: 0;
  }

  @media only screen and (max-width: 700px) {
    border-top-left-radius: 10px;
    border-bottom-left-radius: 10px;
    border-top-right-radius: 10px;
    border-bottom-right-radius: 10px;
  }
`;
const Contents = styled.div`
  width: 50%;
  display: flex;
  flex-direction: column;
`;
const Top = styled.div`
  display: flex;
  flex: 1;
  justify-content: space-between;
  align-items: flex-end;

  @media only screen and (max-width: 700px) {
    justify-content: flex-end;
    align-items: center;
  }
`;
const Bottom = styled.div`
  align-self: flex-end;

  @media only screen and (max-width: 700px) {
    align-self: center;
  }
`;
const Letter = styled.span`
  font-family: ${Fonts.roboto};
  font-size: ${Sizes.variable.font.small};
  color: white;
  text-transform: uppercase;
`;
const LetterSmall = styled(Letter)`
  font-size: ${Sizes.variable.font.xSmall};
`;
const LetterEng = styled(Letter)`
  @media only screen and (max-width: 700px) {
    display: none;
  }
`;

const pressed = { backgroundColor: "#FF7D7D", color: "black", border: "1px solid black" };

function KeyboardRow(props: ({ rowLetters: Array<KeyType>, pressedKey: string })): JSX.Element {
  return (
    <Row numOfEl={props.rowLetters.length}>
      {props.rowLetters.map((key, index) => (
        <Key key={`${key}${index}`} style={props.pressedKey.localeCompare(key.eng) === 0 ? pressed : {}}>
          <Contents>

            <Top>
              <LetterEng>
                {key.eng}
              </LetterEng>
              <LetterSmall>
                {key.shift}
              </LetterSmall>
            </Top>

            <Bottom>
              <Letter>
                {key.kor}
              </Letter>
            </Bottom>

          </Contents>
        </Key>
      ))}
    </Row>
  )
}

interface KeyboardProps extends StyledProps {
  onSubmit: (key: string) => void;
}
function Keyboard(props: KeyboardProps) {
  const [word, setWord] = useState("");
  const [pressedKey, setPressedKey] = useState("");

  function handleChange(e: React.ChangeEvent<HTMLInputElement>): void {
    setPressedKey(e.target.value.substr(-1));
    setWord(e.target.value);
    setTimeout(() => { setPressedKey("") }, 100);
  }

  function handleSubmit(e: React.FormEvent<HTMLFormElement>): void {
    e.preventDefault();
    props.onSubmit(word);
    setPressedKey("");
    setWord("");
  }

  // TODO: update when I update the design for the form input
  return (
    <Container className={props.className}>
      <form onSubmit={handleSubmit}>
        <input
          type="text"
          onChange={handleChange}
          value={word}
          autoFocus />
      </form>
      <KeyboardRow rowLetters={topRow} pressedKey={pressedKey} />
      <KeyboardRow rowLetters={midRow} pressedKey={pressedKey} />
      <KeyboardRow rowLetters={botRow} pressedKey={pressedKey} />
    </Container>
  );
}

export default Keyboard;