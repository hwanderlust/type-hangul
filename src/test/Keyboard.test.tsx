import 'jest-styled-components';
import React from 'react';
import ReactDOM from "react-dom";
import ReactTestUtils from "react-dom/test-utils";
import renderer from 'react-test-renderer';

import Keyboard, { Test } from "../components/common/Keyboard";

const noOp = () => { };

describe("<Keyboard/>", () => {
  let container: HTMLDivElement;
  let onSubmit: (word: string) => void;

  beforeAll(() => {
    container = document.createElement("div");
    onSubmit = jest.fn();
    ReactDOM.render(<Keyboard onSubmit={onSubmit} />, container);
  });

  it('renders correctly', () => {
    const tree = renderer
      .create(<Keyboard onSubmit={noOp} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("doesn't pass up the typed word when typing", () => {
    const input = container.querySelector("input");
    if (input) {
      input.value = "바보";
      ReactTestUtils.Simulate.change(input);
      expect(onSubmit).not.toBeCalled();
    }
  });

  it("passes up the typed word when hitting 'Enter' and input value is cleared", () => {
    const input = container.querySelector("input");
    const form = container.querySelector("form");
    if (input && form) {
      input.value = "바보";
      ReactTestUtils.Simulate.change(input);
      ReactTestUtils.Simulate.submit(form);

      expect(onSubmit).toBeCalledWith("바보");
      expect(input.value).toBe("");
    }
  });
});

describe("Keyboard helpers", () => {
  const { isEnglish, isLowercaseEnglish, isUppercaseEnglish } = Test;

  describe("isLowercaseEnglish()", () => {
    it("detects if a word is a lowercase English letter by the first character only", () => {
      expect(isLowercaseEnglish("a그래")).toBe(true);
      expect(isLowercaseEnglish("gㅋㅋㅋ")).toBe(true);
      expect(isLowercaseEnglish("그래z")).toBe(false);
      expect(isLowercaseEnglish("그123")).toBe(false);
    });
    it("doesn't consider numerical values as 'English'", () => {
      expect(isLowercaseEnglish("123")).toBe(false);
      expect(isLowercaseEnglish("0")).toBe(false);
      expect(isLowercaseEnglish("7asdf")).toBe(false);
      expect(isLowercaseEnglish("666ㅋㅋ")).toBe(false);
    });
  });

  describe("isUppercaseEnglish()", () => {
    it("detects if a word is a uppercase English letter by the first character only", () => {
      expect(isUppercaseEnglish("A그래")).toBe(true);
      expect(isUppercaseEnglish("Gㅋㅋㅋ")).toBe(true);
      expect(isUppercaseEnglish("그래Z")).toBe(false);
      expect(isUppercaseEnglish("그123")).toBe(false);
    });
    it("doesn't consider numerical values as 'English'", () => {
      expect(isUppercaseEnglish("123")).toBe(false);
      expect(isUppercaseEnglish("0")).toBe(false);
      expect(isUppercaseEnglish("7ASDF")).toBe(false);
      expect(isUppercaseEnglish("666ㅋㅋ")).toBe(false);
    });
  });

  describe("isEnglish()", () => {
    it("doesn't error out and returns FALSE if a nullish value", () => {
      expect(isEnglish(undefined)).toBe(false);
      expect(isEnglish(null)).toBe(false);
      expect(isEnglish("")).toBe(false);
    });
    it("detects if a given word is 'English' by its first character only", () => {
      expect(isEnglish("I am a 바보")).toBe(true);
      expect(isEnglish("바보 I am")).toBe(false);
      expect(isEnglish("what is a 바보?")).toBe(true);
      expect(isEnglish("내가!")).toBe(false);
      expect(isEnglish("!!!")).toBe(false);
      expect(isEnglish("! oops")).toBe(false);
    });
  });
});
