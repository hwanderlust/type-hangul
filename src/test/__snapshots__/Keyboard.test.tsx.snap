// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Keyboard/> renders correctly 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  margin-bottom: calc(4px + (8 - 4) * ((100vw - 300px) / (1440 - 300)));
}

.c1:last-of-type {
  margin-bottom: 0;
}

.c2 {
  background-color: black;
  width: calc(25px + (75 - 25) * ((100vw - 300px) / (1440 - 300)));
  height: calc(40px + (75 - 40) * ((100vw - 300px) / (1440 - 300)));
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-top-left-radius: 20px;
  border-bottom-left-radius: 20px;
  border-top-right-radius: 20px;
  border-bottom-right-radius: 20px;
  margin-right: calc(4px + (8 - 4) * ((100vw - 300px) / (1440 - 300)));
}

.c2:last-of-type {
  margin-right: 0;
}

.c3 {
  width: 50%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: flex-end;
  -webkit-box-align: flex-end;
  -ms-flex-align: flex-end;
  align-items: flex-end;
}

.c8 {
  -webkit-align-self: flex-end;
  -ms-flex-item-align: end;
  align-self: flex-end;
}

.c5 {
  font-family: 'Roboto',serif;
  font-size: calc(12px + (16 - 12) * ((100vw - 300px) / (1440 - 300)));
  color: white;
  text-transform: uppercase;
}

.c7 {
  font-size: calc(8px + (12 - 8) * ((100vw - 300px) / (1440 - 300)));
}

@media only screen and (max-width:700px) {
  .c2 {
    border-top-left-radius: 10px;
    border-bottom-left-radius: 10px;
    border-top-right-radius: 10px;
    border-bottom-right-radius: 10px;
  }
}

@media only screen and (max-width:700px) {
  .c4 {
    -webkit-box-pack: end;
    -webkit-justify-content: flex-end;
    -ms-flex-pack: end;
    justify-content: flex-end;
    -webkit-align-items: center;
    -webkit-box-align: center;
    -ms-flex-align: center;
    align-items: center;
  }
}

@media only screen and (max-width:700px) {
  .c8 {
    -webkit-align-self: center;
    -ms-flex-item-align: center;
    align-self: center;
  }
}

@media only screen and (max-width:700px) {
  .c6 {
    display: none;
  }
}

<div
  className="c0"
>
  <form
    onSubmit={[Function]}
  >
    <input
      autoFocus={true}
      onChange={[Function]}
      type="text"
      value=""
    />
  </form>
  <div
    className="c1"
  >
    <div
      className="c2"
    >
      <div
        className="c3"
      >
        <div
          className="c4"
        >
          <span
            className="c5 c6"
          >
            q
          </span>
          <span
            className="c5 c7"
          >
            ㅃ
          </span>
        </div>
        <div
          className="c8"
        >
          <span
            className="c5"
          >
            ㅂ
          </span>
        </div>
      </div>
    </div>
    <div
      className="c2"
    >
      <div
        className="c3"
      >
        <div
          className="c4"
        >
          <span
            className="c5 c6"
          >
            w
          </span>
          <span
            className="c5 c7"
          >
            ㅉ
          </span>
        </div>
        <div
          className="c8"
        >
          <span
            className="c5"
          >
            ㅈ
          </span>
        </div>
      </div>
    </div>
    <div
      className="c2"
    >
      <div
        className="c3"
      >
        <div
          className="c4"
        >
          <span
            className="c5 c6"
          >
            e
          </span>
          <span
            className="c5 c7"
          >
            ㄸ
          </span>
        </div>
        <div
          className="c8"
        >
          <span
            className="c5"
          >
            ㄷ
          </span>
        </div>
      </div>
    </div>
    <div
      className="c2"
    >
      <div
        className="c3"
      >
        <div
          className="c4"
        >
          <span
            className="c5 c6"
          >
            r
          </span>
          <span
            className="c5 c7"
          >
            ㄲ
          </span>
        </div>
        <div
          className="c8"
        >
          <span
            className="c5"
          >
            ㄱ
          </span>
        </div>
      </div>
    </div>
    <div
      className="c2"
    >
      <div
        className="c3"
      >
        <div
          className="c4"
        >
          <span
            className="c5 c6"
          >
            t
          </span>
          <span
            className="c5 c7"
          >
            ㅆ
          </span>
        </div>
        <div
          className="c8"
        >
          <span
            className="c5"
          >
            ㅅ
          </span>
        </div>
      </div>
    </div>
    <div
      className="c2"
    >
      <div
        className="c3"
      >
        <div
          className="c4"
        >
          <span
            className="c5 c6"
          >
            y
          </span>
          <span
            className="c5 c7"
          >
            
          </span>
        </div>
        <div
          className="c8"
        >
          <span
            className="c5"
          >
            ㅛ
          </span>
        </div>
      </div>
    </div>
    <div
      className="c2"
    >
      <div
        className="c3"
      >
        <div
          className="c4"
        >
          <span
            className="c5 c6"
          >
            u
          </span>
          <span
            className="c5 c7"
          >
            
          </span>
        </div>
        <div
          className="c8"
        >
          <span
            className="c5"
          >
            ㅕ
          </span>
        </div>
      </div>
    </div>
    <div
      className="c2"
    >
      <div
        className="c3"
      >
        <div
          className="c4"
        >
          <span
            className="c5 c6"
          >
            i
          </span>
          <span
            className="c5 c7"
          >
            
          </span>
        </div>
        <div
          className="c8"
        >
          <span
            className="c5"
          >
            ㅑ
          </span>
        </div>
      </div>
    </div>
    <div
      className="c2"
    >
      <div
        className="c3"
      >
        <div
          className="c4"
        >
          <span
            className="c5 c6"
          >
            o
          </span>
          <span
            className="c5 c7"
          >
            ㅒ
          </span>
        </div>
        <div
          className="c8"
        >
          <span
            className="c5"
          >
            ㅐ
          </span>
        </div>
      </div>
    </div>
    <div
      className="c2"
    >
      <div
        className="c3"
      >
        <div
          className="c4"
        >
          <span
            className="c5 c6"
          >
            p
          </span>
          <span
            className="c5 c7"
          >
            ㅖ
          </span>
        </div>
        <div
          className="c8"
        >
          <span
            className="c5"
          >
            ㅔ
          </span>
        </div>
      </div>
    </div>
  </div>
  <div
    className="c1"
  >
    <div
      className="c2"
    >
      <div
        className="c3"
      >
        <div
          className="c4"
        >
          <span
            className="c5 c6"
          >
            a
          </span>
          <span
            className="c5 c7"
          >
            
          </span>
        </div>
        <div
          className="c8"
        >
          <span
            className="c5"
          >
            ㅁ
          </span>
        </div>
      </div>
    </div>
    <div
      className="c2"
    >
      <div
        className="c3"
      >
        <div
          className="c4"
        >
          <span
            className="c5 c6"
          >
            s
          </span>
          <span
            className="c5 c7"
          >
            
          </span>
        </div>
        <div
          className="c8"
        >
          <span
            className="c5"
          >
            ㄴ
          </span>
        </div>
      </div>
    </div>
    <div
      className="c2"
    >
      <div
        className="c3"
      >
        <div
          className="c4"
        >
          <span
            className="c5 c6"
          >
            d
          </span>
          <span
            className="c5 c7"
          >
            
          </span>
        </div>
        <div
          className="c8"
        >
          <span
            className="c5"
          >
            ㅇ
          </span>
        </div>
      </div>
    </div>
    <div
      className="c2"
    >
      <div
        className="c3"
      >
        <div
          className="c4"
        >
          <span
            className="c5 c6"
          >
            f
          </span>
          <span
            className="c5 c7"
          >
            
          </span>
        </div>
        <div
          className="c8"
        >
          <span
            className="c5"
          >
            ㄹ
          </span>
        </div>
      </div>
    </div>
    <div
      className="c2"
    >
      <div
        className="c3"
      >
        <div
          className="c4"
        >
          <span
            className="c5 c6"
          >
            g
          </span>
          <span
            className="c5 c7"
          >
            
          </span>
        </div>
        <div
          className="c8"
        >
          <span
            className="c5"
          >
            ㅎ
          </span>
        </div>
      </div>
    </div>
    <div
      className="c2"
    >
      <div
        className="c3"
      >
        <div
          className="c4"
        >
          <span
            className="c5 c6"
          >
            h
          </span>
          <span
            className="c5 c7"
          >
            
          </span>
        </div>
        <div
          className="c8"
        >
          <span
            className="c5"
          >
            ㅗ
          </span>
        </div>
      </div>
    </div>
    <div
      className="c2"
    >
      <div
        className="c3"
      >
        <div
          className="c4"
        >
          <span
            className="c5 c6"
          >
            j
          </span>
          <span
            className="c5 c7"
          >
            
          </span>
        </div>
        <div
          className="c8"
        >
          <span
            className="c5"
          >
            ㅓ
          </span>
        </div>
      </div>
    </div>
    <div
      className="c2"
    >
      <div
        className="c3"
      >
        <div
          className="c4"
        >
          <span
            className="c5 c6"
          >
            k
          </span>
          <span
            className="c5 c7"
          >
            
          </span>
        </div>
        <div
          className="c8"
        >
          <span
            className="c5"
          >
            ㅏ
          </span>
        </div>
      </div>
    </div>
    <div
      className="c2"
    >
      <div
        className="c3"
      >
        <div
          className="c4"
        >
          <span
            className="c5 c6"
          >
            l
          </span>
          <span
            className="c5 c7"
          >
            
          </span>
        </div>
        <div
          className="c8"
        >
          <span
            className="c5"
          >
            ㅣ
          </span>
        </div>
      </div>
    </div>
  </div>
  <div
    className="c1"
  >
    <div
      className="c2"
    >
      <div
        className="c3"
      >
        <div
          className="c4"
        >
          <span
            className="c5 c6"
          >
            z
          </span>
          <span
            className="c5 c7"
          >
            
          </span>
        </div>
        <div
          className="c8"
        >
          <span
            className="c5"
          >
            ㅋ
          </span>
        </div>
      </div>
    </div>
    <div
      className="c2"
    >
      <div
        className="c3"
      >
        <div
          className="c4"
        >
          <span
            className="c5 c6"
          >
            x
          </span>
          <span
            className="c5 c7"
          >
            
          </span>
        </div>
        <div
          className="c8"
        >
          <span
            className="c5"
          >
            ㅌ
          </span>
        </div>
      </div>
    </div>
    <div
      className="c2"
    >
      <div
        className="c3"
      >
        <div
          className="c4"
        >
          <span
            className="c5 c6"
          >
            c
          </span>
          <span
            className="c5 c7"
          >
            
          </span>
        </div>
        <div
          className="c8"
        >
          <span
            className="c5"
          >
            ㅊ
          </span>
        </div>
      </div>
    </div>
    <div
      className="c2"
    >
      <div
        className="c3"
      >
        <div
          className="c4"
        >
          <span
            className="c5 c6"
          >
            v
          </span>
          <span
            className="c5 c7"
          >
            
          </span>
        </div>
        <div
          className="c8"
        >
          <span
            className="c5"
          >
            ㅍ
          </span>
        </div>
      </div>
    </div>
    <div
      className="c2"
    >
      <div
        className="c3"
      >
        <div
          className="c4"
        >
          <span
            className="c5 c6"
          >
            b
          </span>
          <span
            className="c5 c7"
          >
            
          </span>
        </div>
        <div
          className="c8"
        >
          <span
            className="c5"
          >
            ㅠ
          </span>
        </div>
      </div>
    </div>
    <div
      className="c2"
    >
      <div
        className="c3"
      >
        <div
          className="c4"
        >
          <span
            className="c5 c6"
          >
            n
          </span>
          <span
            className="c5 c7"
          >
            
          </span>
        </div>
        <div
          className="c8"
        >
          <span
            className="c5"
          >
            ㅜ
          </span>
        </div>
      </div>
    </div>
    <div
      className="c2"
    >
      <div
        className="c3"
      >
        <div
          className="c4"
        >
          <span
            className="c5 c6"
          >
            m
          </span>
          <span
            className="c5 c7"
          >
            
          </span>
        </div>
        <div
          className="c8"
        >
          <span
            className="c5"
          >
            ㅡ
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;
